src/consensus/tower.zig

```diff
@@ -1982,136 +1470,729 @@ pub const Tower = struct {
+pub fn lastVotedSlotInBank(
+    allocator: std.mem.Allocator,
+    accounts_db: *AccountsDB,
+    vote_account_pubkey: *const Pubkey,
+) ?Slot {
+    const vote_account = accounts_db.getAccount(vote_account_pubkey) catch return null;
+    const vote_state = stateFromAccount(
+        allocator,
+        &vote_account,
+        vote_account_pubkey,
+    ) catch return null;
+    return vote_state.lastVotedSlot();
+}
 
-        var tower = try createTestTower(allocator, 8, 0.66);
-        defer tower.deinit(allocator);
+fn stateFromAccount(
+    allocator: std.mem.Allocator,
+    vote_account: *const Account,
+    vote_account_pubkey: *const Pubkey,
+) !VoteState {
+    const buf = try allocator.alloc(u8, vote_account.data.len());
+    // TODO Not sure if this is the way to get the data from the vote account. Review.
+    _ = vote_account.writeToBuf(vote_account_pubkey, buf);
+    const versioned_state = try sig.bincode.readFromSlice(
+        allocator,
+        VoteStateVersions,
+        buf,
+        .{},
+    );
+    return try versioned_state.convertToCurrent(allocator);
+}
 
-        const latest = LatestValidatorVotesForFrozenBanks{
-            .max_gossip_frozen_votes = std.AutoHashMap(Pubkey, FrozenVotes).init(allocator),
-        };
+/// Checks if `maybe_descendant` is a descendant of `slot`.
+///
+/// Returns none if `maybe_descendant` is not present in `ancestors`
+fn isDescendantSlot(
+    maybe_descendant: Slot,
+    slot: Slot,
+    ancestors: *const AutoHashMapUnmanaged(Slot, SortedSet(Slot)),
+) ?bool {
+    return if (ancestors.get(maybe_descendant)) |candidate_slot_ancestors|
+        candidate_slot_ancestors.contains(slot)
+    else
+        null;
+}
 
-        var slot_history = try createTestSlotHistory(std.testing.allocator);
-        defer slot_history.deinit(allocator);
+/// Returns `Some(gca)` where `gca` is the greatest (by slot number)
+/// common ancestor of both `slot_a` and `slot_b`.
+///
+/// Returns `null` if:
+/// * `slot_a` is not in `ancestors`
+/// * `slot_b` is not in `ancestors`
+/// * There is no common ancestor of slot_a and slot_b in `ancestors`
+fn greatestCommonAncestor(
+    ancestors: *const AutoHashMapUnmanaged(
+        Slot,
+        SortedSet(Slot),
+    ),
+    slot_a: Slot,
+    slot_b: Slot,
+) ?Slot {
+    var ancestors_a = ancestors.get(slot_a) orelse return null;
+    var ancestors_b = ancestors.get(slot_b) orelse return null;
 
-        try std.testing.expectError(
-            error.StakeNotFound,
-            tower.selectVoteAndResetForks(
-                std.testing.allocator,
-                100,
-                null,
-                100,
-                &.{},
-                &.{},
-                &ProgressMap.INIT,
-                &latest,
-                &fork_choice,
-                .{},
-                &slot_history,
-            ),
-        );
+    var max_slot: ?Slot = null;
+
+    var superset, const subset = if (ancestors_a.count() >= ancestors_b.count())
```

Comment:
Should `const`ify some of these vars

Reply:
[2e2ebce](https://github.com/Syndica/sig/pull/724/commits/2e2ebce2b083c0bab7074c204a36b93212c68053)